src/client/reconnecting_websocket.coffee

```diff
@@ -144,27 +108,100 @@ class ReconnectingWebSocket
       @forcedClose = true
       @ws.close()
 
-  ###
-  Setting this to true is the equivalent of setting all instances of ReconnectingWebSocket.debug to true.
-  ###
-  debugAll: false
   ###
   Additional public API method to refresh the connection if still open (close, re-open).
   For example, if the app suspects bad data / missed heart beats, it can try to refresh.
   ###
   refresh: ->
-    @ws.close()  if @ws
+    console.log "ReconnectingWebSocket", "refresh" if @debug
+    @_reconnect()
 
-  # This checks periodically to see if the server is still responding.
+  # Closes the WebSocket connection or connection attempt and tries again.
+  #
+  # Useful if stuck in a state unexpectedly, or if no heartbeat response
+  # has been received.
+  _reconnect: =>
+    console.log "ReconnectingWebSocket", "reconnect"  if @debug
+    if !@ws || @ws.readyState == WebSocket.CLOSED
+      @_connect()
+    else
+      @_disconnect()
+      @_connect()
+
+  # This checks to see if the server is still responding.
   #
   # If the client does not receive a response within the interval, it
   # will refresh the connection.
-  _periodicHeartbeatCheck: ->
-    @heartbeat = true
-    @checkInterval = setInterval =>
-      if @heartbeat
-        @heartbeat = null
-        @send JSON.stringify "heartbeat"
-      else
-        @refresh()
-    , HEARTBEAT_INTERVAL
+  _checkHeartbeat: =>
+    if @heartbeatResponse == HEARTBEAT_REQUESTED
+      console.log "ReconnectingWebSocket", "no-heartbeat"  if @debug
+      @_reconnect()
+    else
+      @heartbeatResponse = HEARTBEAT_REQUESTED
+      console.log "ReconnectingWebSocket", "send-heartbeat"  if @debug
+      @send JSON.stringify "heartbeat"
+
+  _periodicHealthCheck: =>
+    console.log "ReconnectingWebSocket", "healthcheck"  if @debug
+    switch @readyState
+      when WebSocket.CLOSED then @_reconnect()
+      when WebSocket.CONNECTING then @_reconnect()
+      when WebSocket.OPEN then @_checkHeartbeat()
+
+  _connect: =>
+    @ws = new WebSocket(@url)
+    @readyState = WebSocket.CONNECTING
+    console.log "ReconnectingWebSocket", "attempt-connect", @url  if @debug
+    @ws.addEventListener "open", @_handleWebsocketOpen
+    @ws.addEventListener "close", @_handleWebsocketClose
+    @ws.addEventListener "message", @_handleWebsocketMessage
+    @ws.addEventListener "error", @_handleWebsocketError
```

Comment:
Actually creating a new connection is way better than what was happening when trying to rescue the dropped one! 👏